python/ray/data/preprocessor.py

```diff
@@ -226,59 +254,83 @@ def _transform(self, dataset: Dataset) -> Dataset:
             )
 
         transform_type = self._determine_transform_to_use(dataset_format)
 
+        # Our user facing batch format should only be pandas or numpy, other
+        # formats {arrow, simple} are internal.
         if transform_type == "pandas":
             return dataset.map_batches(self._transform_pandas, batch_format="pandas")
-        elif transform_type == "arrow":
-            return dataset.map_batches(self._transform_arrow, batch_format="pyarrow")
+        elif transform_type == "numpy":
+            return dataset.map_batches(self._transform_numpy, batch_format="numpy")
         else:
             raise ValueError(
                 "Invalid transform type returned from _determine_transform_to_use; "
-                f'"pandas" and "arrow" allowed, but got: {transform_type}'
+                f'"pandas" and "numpy" allowed, but got: {transform_type}'
             )
 
-    def _transform_batch(self, df: "DataBatchType") -> "DataBatchType":
+    def _transform_batch(self, data: "DataBatchType") -> "DataBatchType":
         import pandas as pd
 
         try:
             import pyarrow
         except ImportError:
             pyarrow = None
 
-        if isinstance(df, pd.DataFrame):
+        if isinstance(data, pd.DataFrame):
             data_format = "pandas"
-        elif pyarrow is not None and isinstance(df, pyarrow.Table):
+        elif pyarrow is not None and isinstance(data, pyarrow.Table):
             data_format = "arrow"
+        elif isinstance(data, (dict, np.ndarray)):
+            data_format = "numpy"
         else:
             raise NotImplementedError(
-                "`transform_batch` is currently only implemented for Pandas DataFrames "
-                f"and PyArrow Tables. Got {type(df)}."
+                "`transform_batch` is currently only implemented for Pandas "
+                "DataFrames, PyArrow Tables, Numpy ndarray and dictionary of "
+                f"ndarary. Got {type(data)}."
             )
 
         transform_type = self._determine_transform_to_use(data_format)
 
         if transform_type == "pandas":
             if data_format == "pandas":
-                return self._transform_pandas(df)
-            else:
-                return self._transform_pandas(df.to_pandas())
-        elif transform_type == "arrow":
-            if data_format == "arrow":
-                return self._transform_arrow(df)
+                return self._transform_pandas(data)
             else:
-                return self._transform_arrow(pyarrow.Table.from_pandas(df))
+                return self._transform_pandas(data.to_pandas())
+        elif transform_type == "numpy":
+            if data_format == "numpy":
+                return self._transform_numpy(data)
+            elif data_format == "arrow":
+                if len(data.column_names) == 1:
+                    # If just a single column, return as a single numpy array.
+                    return self._transform_numpy(data[0].to_numpy())
```

Comment:
Hmm I think that single-column tables will need to be preserved as tables unless it's using the Datasets scheme for representing "tensor datasets", i.e. a single `"__value__"` column that's using our tensor extension dtype.

Something like:
```suggestion
                from ray.data._internal.table_block import VALUE_COL_NAME
                from ray.data.extensions.tensor_extension import ArrowTensorType
                
                if data.column_names == [VALUE_COL_NAME] and isinstance(data.schema.types[0], ArrowTensorType):
                    # If representing a tensor dataset, return as a single numpy array.
                    return self._transform_numpy(data[0].to_numpy())
```

Reply:
Question on this, don't we already have a utility for converting blocks to our desired batch format? Can we call that utility instead of re-encoding the logic in preprocessor?

Reply:
hmm is this the api you're mentioning ? 

`BlockAccessor.for_block(view).to_batch_format(batch_format)` 

looks like it operates at block level where preprocessor appears to be data batch level with `DataBatchType = Union[
    "numpy.ndarray", "pandas.DataFrame", "pyarrow.Table", Dict[str, "numpy.ndarray"]
]`

But if we have a single and consistent API for data format -> block format -> batch format conversion that will be great, not sure if we already had one in place.

Reply:
A better option that makes the batch <--> block interop well-defined would be `BlockAccessor.batch_to_block(data).to_batch_format(batch_format)`, but this will convert `Union[np.ndarray, Dict[str, np.ndarray]]` to an Arrow Table, possibly unnecessarily if `batch_format` ends up being `"numpy"`. This _should_ usually be zero-copy for the underlying ndarray buffers, but I know that this will at least cause a column copy in the following cases:
1. Boolean column (Arrow bitpacks boolean arrays while NumPy does not).
2. Ragged tensor after upcoming PR (Arrow representation is contiguous in memory while NumPy representation is an ndarray of ndarray pointers).
3. Not C-contiguous, e.g. Fortran order.
4. String column (Arrow allocates these as contiguous string arrays while NumPy has them as ndarray of string pointers).

Reply:
I think @ericl is referring to these: https://github.com/ray-project/ray/blob/master/python/ray/air/util/data_batch_conversion.py

which we use for pandas narrow waist. We should extend these utils to also support numpy narrow waist. 

We shouldn't be using any `BlockAccessor` code directly in `_transform_batch`

Reply:
Having the conversions in util functions means we can leverage them for Predictor numpy narrow waist as well

Reply:
i see, this is a good suggestion. We should keep this layer of conversion in one place.

Reply:
Ah yep +1 to the AIR-level `data_batch_conversion.py` utilities, I forgot about those! 😅